File: day12.py
"""Solution to day 12 of Advent of Code"""

from get_input import get_input, line_parser


def move(direction):
    def move_func(heading, position, value):
        return heading, position + (direction * value)
    return move_func


def turn(direction):
    def turn_func(heading, position, value):
        while value > 0:
            heading = direction(heading)
            value -= 90
        assert value == 0
        return heading, position
    return turn_func


RULES = {
    'N': move(0+1j),
    'S': move(0-1j),
    'E': move(1+0j),
    'W': move(-1+0j),
    'R': turn(lambda h: complex(h.imag, -h.real)),
    'L': turn(lambda h: complex(-h.imag, h.real)),
    'F': lambda h, p, v: (h, p + h * v),
}


def part1(lines):
    heading = 1+0j
    position = 0+0j
    rules = RULES.copy()
    for (order, value) in lines:
        heading, position = rules[order](heading, position, value)
    return int(abs(position.real) + abs(position.imag))


def swap(function):
    def swapped(heading, position, value):
        position, heading = function(position, heading, value)
        return heading, position
    return swapped


def part2(lines):
    waypoint = 10+1j
    position = 0+0j
    rules = RULES.copy()
    for order in ('N', 'S', 'E', 'W'):
        rules[order] = swap(rules[order])
    for (order, value) in lines:
        waypoint, position = rules[order](waypoint, position, value)
    return int(abs(position.real) + abs(position.imag))


def parse(line):
    return (line[0], int(line[1:]))


TEST = """F10
N3
F7
R90
F11"""

if __name__ == "__main__":
    assert part1(line_parser(TEST, parse=parse)) == 25
    assert part2(line_parser(TEST, parse=parse)) == 286
    LINES = line_parser(get_input(day=12, year=2020), parse=parse)
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day12.rs
use common::get_input;

type Input = Vec<(String, usize)>;
type Point = (isize, isize);
type Update = Box<dyn Fn(Point, Point, &usize) -> (Point, Point)>;

fn move_point(diff: Point) -> Update {
    Box::new(move |heading, position, &value| {
        (
            heading,
            (position.0 + diff.0 * (value as isize), position.1 + diff.1 * (value as isize))
        )
    })
}

fn turn(update: Box<dyn Fn(Point) -> Point>) -> Update {
    Box::new(move |heading, position, &value| {
        let mut heading = heading;
        let mut value = value;
        while value > 0 {
            value -= 90;
            heading = update(heading);
        }
        assert!(value == 0);
        (heading, position)
    })
}

fn part1(directions: &Input) -> isize {
    let mut heading = (1, 0);
    let mut position = (0, 0);
    for (order, value) in directions {
        let forward: Update = Box::new(|heading, position, &value| {
            (heading, (position.0 + value as isize * heading.0, position.1 + value as isize * heading.1))
        });
        let update = match order.as_str() {
            "N" => move_point((0, 1)),
            "S" => move_point((0, -1)),
            "E" => move_point((1, 0)),
            "W" => move_point((-1, 0)),
            "R" => turn(Box::new(|(r, i)| (i, -r))),
            "L" => turn(Box::new(|(r, i)| (-i, r))),
            "F" => forward,
            _ => unimplemented!(),
        };
        let result = update(heading, position, value);
        heading = result.0;
        position = result.1;
    }
    position.0.abs() + position.1.abs()
}

fn swap(func: Update) -> Update {
    Box::new(move |heading, position, value| {
        let result = func(position, heading, value);
        (result.1, result.0)
    })
}

fn part2(directions: &Input) -> isize {
    let mut waypoint = (10, 1);
    let mut position = (0, 0);
    for (order, value) in directions {
        let forward: Update = Box::new(|heading, position, &value| {
            (heading, (position.0 + value as isize * heading.0, position.1 + value as isize * heading.1))
        });
        let update = match order.as_str() {
            "N" => swap(move_point((0, 1))),
            "S" => swap(move_point((0, -1))),
            "E" => swap(move_point((1, 0))),
            "W" => swap(move_point((-1, 0))),
            "R" => turn(Box::new(|(r, i)| (i, -r))),
            "L" => turn(Box::new(|(r, i)| (-i, r))),
            "F" => forward,
            _ => unimplemented!(),
        };
        let result = update(waypoint, position, value);
        waypoint = result.0;
        position = result.1;
    }
    position.0.abs() + position.1.abs()
}

fn parse(text: String) -> Input {
    text.trim()
        .split('\n')
        .map(|line| {
            (
                line.chars().next().unwrap().to_string(),
                line.chars().skip(1).collect::<String>().parse::<usize>().unwrap(),
            )
        })
        .collect()
}

fn main() {
    let input = parse(get_input(12, 2020)); 
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}

